### PR TITLE
simplify (show, URL)

### PR DESCRIPTION
**Pull request by mahrud**
This PR simplifies part of `html.m2` that deals with opening a browser. With this change, instead of specifying how to open a few browsers, we look for the `open` on macOS and `[xdg-open](https://www.freedesktop.org/wiki/Software/xdg-utils/)` on Linux. I'm fairly sure all Linux distributions that we support come with `xdg-utils` scripts, but I haven't tested them all, so as a backup we still check if firefox exists.

### CODE REVIEW

**Comment by DanGrayson** [comment]
The environment variable WWWBROWSER is most likely documented and users might be depending on it, so it should still be referred to first.  

Could you also update the documentation to refer to xdg-open (which I wasn't aware of)?  And give a hint about how one might set the browser preference?

**Comment by mahrud** [comment]
I can't find any documentation for `WWWBROWSER` other than two mentions in the google groups a few years ago, but I added info about open/xdg-open. Since it's also not documented, can we deprecate it? We can't anticipate all the browsers someone might want to use in `browserMethods` and if we use `WWWBROWSER` we also have to keep `browserMethods` since just running whatever is in an environment variable isn't safe.

**Comment by DanGrayson** [comment]
We don't have to anticipate all the browsers, because the default is to run the command with one argument consisting of the file name.

I can't find the documentation, either, but I bet people are using it.  It's needed for the case where xdg-open and open are not there, so we can't get rid of it.  

Another point about 'xdg-open' is that the preference probably must be set on the gnome or kde desktop, and not everyone uses a desktop GUI, either.

**Comment by mahrud** [comment]
> We don't have to anticipate all the browsers, because the default is to run the command with one argument consisting of the file name.

I think this unsafe in general because `WWWBROWSER` could set to something dangerous. If we don't mind this, why not let users set wwwbrowser in `init.m2` instead?

> Another point about 'xdg-open' is that the preference probably must be set on the gnome or kde desktop, and not everyone uses a desktop GUI, either.

It's set by browsers when they ask you to make them the default browser. In particular, I tested xdg-open in a headless remote shell and Links 2.12 popper up! So I still think this is a preferred standard over an environment variable, but I added it again.

**Comment by mahrud** [comment]
Just realized there was an extra semicolon in the last commit that made it fail. It's resolved now.

**Comment by DanGrayson** [comment]
> > We don't have to anticipate all the browsers, because the default is to run the command with one argument consisting of the file name.
> 
> I think this unsafe in general because `WWWBROWSER` could set to something dangerous. If we don't mind this, why not let users set wwwbrowser in `init.m2` instead?

Why is "init.m2" any safer?  It could also run something dangerous.

**Comment by mahrud** [comment]
Absolutely. I meant if we don't mind that danger, setting it in init.m2 is easier than setting an environment variable before running M2.

**Comment by DanGrayson** [comment]
I think we don't have a way to set it in `init.m2`.

**Comment by mahrud** [comment]
Is there anything else needed on this PR?

File: M2/Macaulay2/m2/html.m2 [hunk 1]
```diff
@@ -1181,38 +1181,20 @@ makePackageIndex List := path -> ( -- TO DO : rewrite this function to use the r
...
-	  if runnable "firefox" then browser = "firefox"
-	  else if runnable "netscape" then browser = "netscape"
-	  else error "no browser found, and none specified in $WWWBROWSER"
-	  );
-     cmd := if browserMethods#?browser then browserMethods#browser url else { browser, url };
+     if runnable "open" then browser := "open" -- Apple varieties
+     else if runnable "xdg-open" then browser = "xdg-open" -- most Linux distributions
+     else if runnable getenv "WWWBROWSER" then browser = getenv "WWWBROWSER" -- compatibility
+     else if runnable "firefox" then browser = "firefox" -- backup
+     else error "neither open nor xdg-open are found and WWWBROWSER is not set";
```
Comment (DanGrayson) [inline M2/Macaulay2/m2/html.m2:1196]: The error message could also mention that firefox was not found.

File: M2/Macaulay2/packages/Macaulay2Doc/doc1.m2 [hunk 2]
```diff
@@ -94,9 +94,13 @@ document {
 	  "s" => "a descriptor for a documentation node (see below for examples)"
 	  },
      Consequences => {
-	  {"The given documentation page is displayed in your default web browser.  If
-	  the browser is not running, it is started.  If no argument is given to ", TT "viewHelp", 
-	  "then the top page of your local html documentation is displayed."}},
+	  {"The given documentation page is displayed in your default web browser, as determined
+	    by either ", TT "open", " on macOS or ", TT "xdg-open", " on Linux distributions.
+	    As backup for when neither ", TT "open", " nor ", TT "xdg-open", " are available,
+	    the environmental variable ", TT "WWWBROWSER", " or ", TT "firefox", " are used.
```
Comment (DanGrayson) [inline M2/Macaulay2/packages/Macaulay2Doc/doc1.m2:100]: Change "are" to "is".  On the previous line, also.

**Comment by DanGrayson** [comment]
I've added a couple of comments.  Also, I've noticed that if the browser cannot be found, then
this code does not flag an error in the parent process:
```
     if fork() == 0 then (
	  setGroupID(0,0);
     	  try exec cmd;
     	  stderr << "exec failed: " << toExternalString cmd << endl;
	  exit 1
	  )
```
Can you think of a better way to do it?

**Comment by DanGrayson** [comment]
What about adding "chrome" to the list?  (See #1152 , where I was not remembering this pull request.)

**Comment by mahrud** [comment]
I think we should not add individual browsers and instead use the OS to decide that for us, based on the users preferences. I wanted to remove Firefox as well, but left it because you asked and for backwards compatibility. Even if not for that reason, the binary of chrome is `google-chrome` on one of my machines and `chromium` on another.

**Comment by mahrud** [comment]
> I've added a couple of comments. Also, I've noticed that if the browser cannot be found, then
> this code does not flag an error in the parent process:
> 
> ```
>      if fork() == 0 then (
> 	  setGroupID(0,0);
>      	  try exec cmd;
>      	  stderr << "exec failed: " << toExternalString cmd << endl;
> 	  exit 1
> 	  )
> ```
> 
> Can you think of a better way to do it?

I personally prefer that it didn't flag an error. If you'd like to push a commit to this PR that raises an error or reverts to "help", feel free to. I've allowed commits to this PR from maintainers.

**Comment by mahrud** [comment]
I'm working on a larger refactoring in the Core that would conflict with this if it isn't merged soon. If you think more work is left, I'm inclined to close this and solve the issue separately.

**Pull request merged**